bisection.go

```diff
@@ -57,7 +57,20 @@ func (b *Bisection) Init(loc LinesearchLocation, step float64, _ *FunctionInfo)
 }
 
 func (b *Bisection) Finished(loc LinesearchLocation) bool {
-	return StrongWolfeConditionsMet(loc.F, loc.Derivative, b.initF, b.initGrad, b.currStep, 0, b.GradConst)
+	// Don't finish the linesearch until a minimum is found that is better than
+	// the best point found so far. We want to end up in the lowest basin of
+	// attraction
+	minF := b.initF
+	if b.maxF < b.initF {
+		minF = b.maxF
+	}
+	if b.minF < minF {
+		minF = b.minF
+	}
+	if StrongWolfeConditionsMet(loc.F, loc.Derivative, minF, b.initGrad, b.currStep, 0, b.GradConst) {
```

Comment:
To add, in particular, by testing with min_f rather than init_f, when we find the minimum on the left, we don't want to converge -- we want to find the minimum on the right. The linesearch knows this minimum exists because we have something like
````
x_init = 0, f_init = 10, g_init = -1
x_left = 1, f_left = 8, g_left = 0.1
x_right = 2, f_right = 5, g_right = 3
````
so there must be a local minimum and maximum in order to get from x_left to x_right. When we see the function value is higher at x_left, instead of stopping because it's the first point we've found that meets Wolfe, we decide to find that local minimum we know must exist near x_right, and so we choose to bisect between x_left and x_right.